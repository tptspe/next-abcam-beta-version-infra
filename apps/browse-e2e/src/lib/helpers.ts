export const handleException = () => {
  cy.on('uncaught:exception', function (error) {
    if (error.message.includes("Cannot set property 'status' of undefined")) {
      return false
    }
  })
}
