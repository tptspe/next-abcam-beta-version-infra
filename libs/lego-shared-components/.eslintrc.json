{
  "extends": [
    "plugin:@nrwl/nx/react",
    "../../.eslintrc.json",
    "plugin:jsx-a11y/recommended",
    "plugin:cypress/recommended"
  ],
  "ignorePatterns": ["!**/*", "*.stories.*", "*.spec.*", ".storybook/*"],
  "plugins": ["no-relative-import-paths"],
  "rules": {
    "no-relative-import-paths/no-relative-import-paths": [
      "warn",
      { "allowSameFolder": true }
    ]
  },
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": ["libs/lego-shared-components/tsconfig.*?.json"]
      },
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    }
  ]
}
