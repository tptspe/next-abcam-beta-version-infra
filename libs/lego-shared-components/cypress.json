{
  "fileServerFolder": ".",
  "fixturesFolder": "./e2e/fixtures",
  "integrationFolder": "./e2e/integration",
  "modifyObstructiveCode": false,
  "pluginsFile": "./e2e/plugins/index",
  "supportFile": "./e2e/support/index.ts",
  "video": true,
  "videosFolder": "../../dist/cypress/libs/lego-shared-components/videos",
  "screenshotsFolder": "./e2e/screenshots",
  "screenshotOnRunFailure": false,
  "videoCompression": false,
  "chromeWebSecurity": false
}
