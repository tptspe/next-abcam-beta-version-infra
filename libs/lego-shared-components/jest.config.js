module.exports = {
  displayName: 'lego-shared-components',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  moduleNameMapper: {
    '\\.svg$': '<rootDir>/mocks/svg-loader.js',
  },
  coverageDirectory: '../../coverage/libs/lego-shared-components',
}
