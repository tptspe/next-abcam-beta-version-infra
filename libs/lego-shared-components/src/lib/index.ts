export { Button } from './button/button'
export { Select } from './select/select'
export { Dropdown } from './dropdown/dropdown'
