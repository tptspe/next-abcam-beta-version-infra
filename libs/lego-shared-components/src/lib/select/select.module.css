.select {
  @apply text-body-small;
  @apply font-sb;
  @apply relative;
  @apply inline-block;
  @apply cursor-pointer;

  .selectedValueContainer {
    @apply bg-grey95;
    @apply text-grey20;
    @apply flex;
    @apply flex-row;
    @apply justify-center;
    @apply items-center;
    @apply gap-2;
    @apply py-2.5;
    @apply px-4;
    @apply rounded-4px;

    &:focus {
      @apply outline-none;
      @apply shadow-interactiveElement;
    }

    svg {
      @apply w-5;
      @apply h-5;
    }

    &:hover {
      @apply bg-zircon;
    }
  }

  .content {
    @apply absolute;
    @apply top-0;
    @apply z-10;
    @apply w-full;
    @apply cursor-pointer;

    ul {
      @apply text-grey20;
      @apply bg-grey95;
      @apply shadow-elevation-2;
      @apply overflow-scroll;

      max-height: 26.625rem;

      li {
        @apply py-1;

        div {
          @apply pl-3;
          @apply py-2;
          @apply flex;
          margin-left: 0.19rem;
          margin-right: 0.19rem;

          &:hover {
            @apply bg-zircon;
          }
          &:focus {
            @apply outline-none;
            @apply shadow-interactiveElement;
          }
        }
      }
    }
  }
}

/****** Open state ******/
.open {
  .selectedValueContainer {
    @apply rounded-t-4px;
  }
  .content {
    @apply block;
  }
}
