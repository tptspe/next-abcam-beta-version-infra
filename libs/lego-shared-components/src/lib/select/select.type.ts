type SelectOption = { key: string | number; displayValue: string }

type SelectProps = {
  options: SelectOption[]
  onChange: (val: SelectOption) => void
}

export type { SelectProps, SelectOption }
