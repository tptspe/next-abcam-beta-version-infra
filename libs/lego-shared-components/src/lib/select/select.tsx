import React, { FC, useCallback, useRef, useState } from 'react'
import classnames from 'classnames'

import styles from './select.module.css'

import { SelectProps } from './select.type'
import { useClickOutside } from '@abcam-web/lego-shared-components/lib/utils/useClickOutside'
import { ChevronDown } from '@abcam-web/lego-shared-components/icons'
import { ListItem } from '@abcam-web/lego-shared-components/lib/listItem/listItem'
import {
  isArrowDownEvent,
  isEnterEvent,
  isSpaceEvent,
  keyboardVerticalNavigationHandler,
} from '@abcam-web/lego-shared-components/lib/utils/keyboardEvents'

const Select: FC<SelectProps> = (props) => {
  const containerRef = useRef(null)
  const [selectedItemIndex, setSelectedItemIndex] = useState(0)

  const { options, onChange } = props

  const [initialValue] = options
  const [selectedValue, setSelectedValue] = useState(initialValue)
  const [isVisible, setIsVisible] = useState(false)

  const toggleIsVisible = useCallback(() => {
    setIsVisible(!isVisible)
  }, [isVisible, setIsVisible])

  const hideOnOutsideClick = useCallback(() => {
    if (isVisible) {
      setIsVisible(false)
    }
  }, [isVisible, setIsVisible])

  useClickOutside(containerRef, hideOnOutsideClick)

  const setSelectedValueCallback = useCallback(
    (value) => {
      setSelectedValue(value)
      toggleIsVisible()
      onChange(value)
      setSelectedItemIndex(0)
    },
    [setSelectedValue, toggleIsVisible, setSelectedItemIndex, onChange]
  )

  const listKeyUpHandler = useCallback(
    (evt) => {
      if (isEnterEvent(evt) || isSpaceEvent(evt) || isArrowDownEvent(evt)) {
        toggleIsVisible()
      }
    },
    [toggleIsVisible]
  )

  const listDropdownKeyUpHandler = useCallback(
    (evt) =>
      keyboardVerticalNavigationHandler(
        evt,
        selectedItemIndex,
        setSelectedItemIndex,
        options.length - 1
      ),
    [selectedItemIndex, setSelectedItemIndex, options]
  )

  const listItemKeyUpHandler = useCallback(
    (evt, value) => {
      if (isEnterEvent(evt)) {
        setSelectedValueCallback(value)
      }
    },
    [setSelectedValueCallback]
  )

  const classes = classnames({
    [styles.select]: true,
    [styles.open]: isVisible,
  })

  return (
    <div className={classes} ref={containerRef}>
      <div
        tabIndex={0}
        className={styles.selectedValueContainer}
        role="button"
        onClick={toggleIsVisible}
        onKeyUp={listKeyUpHandler}
      >
        {selectedValue.displayValue}
        <ChevronDown aria-label="Open select" />
      </div>
      <div className={styles.content} role={'button'}>
        {isVisible && (
          <ul onKeyUp={listDropdownKeyUpHandler} role="presentation">
            {options.map((val, idx) => {
              return (
                <ListItem
                  autofocus={idx === selectedItemIndex}
                  clickCallback={setSelectedValueCallback}
                  keyupCallback={listItemKeyUpHandler}
                  value={val}
                  key={val.key}
                />
              )
            })}
          </ul>
        )}
      </div>
    </div>
  )
}

export { Select }
