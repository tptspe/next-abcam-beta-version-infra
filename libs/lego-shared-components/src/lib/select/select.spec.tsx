import { screen, fireEvent, render } from '@testing-library/react'
import React from 'react'
import { SelectOption } from './select.type'
import userEvent from '@testing-library/user-event'
import { Select } from '@abcam-web/lego-shared-components/lib'

const numberRange: SelectOption[] = []
for (let i = 1; i <= 10; i++) {
  numberRange.push({ key: i, displayValue: `value ${i}` })
}

describe('The select element', () => {
  describe('Snapshots', () => {
    it('Matches the snapshot on initial state', () => {
      const { container } = render(
        <Select options={numberRange} onChange={() => true} />
      )
      expect(container.firstChild).toMatchSnapshot()
    })

    it('Matches the snapshot when opened', () => {
      const { container } = render(
        <Select options={numberRange} onChange={() => true} />
      )
      const clickElement = screen.getByText('value 1')
      fireEvent.click(clickElement)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('Before the select is clicked', () => {
    it('Does not display the list of values', () => {
      render(<Select options={numberRange} onChange={() => true} />)

      const listItems = screen.queryByRole('listitem')
      expect(listItems).toBeNull()
    })
  })

  describe('When the select is clicked', () => {
    it('Displays the list of values', () => {
      render(<Select options={numberRange} onChange={() => true} />)

      const clickElement = screen.getByText('value 1')
      userEvent.click(clickElement)

      const itemList = screen.getAllByRole('listitem')
      expect(itemList).toHaveLength(numberRange.length)
    })
  })

  describe('When the value is changed', () => {
    it("Passes the selected value to the parent's callback", () => {
      const onChangeSpy = jest.fn()
      render(<Select options={numberRange} onChange={onChangeSpy} />)

      const clickElement = screen.getByText('value 1')
      userEvent.click(clickElement)

      const listItem = screen.getByText('value 3')
      userEvent.click(listItem)

      expect(onChangeSpy).toHaveBeenCalledWith(numberRange[2])
    })
  })

  describe('The keyboard interactions', () => {
    describe('When the select is closed', () => {
      it('Gets into focus when the user presses tab', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')

        spyOn(clickElement, 'focus')

        userEvent.tab()

        expect(clickElement.focus).toHaveBeenCalled()
      })
      it('Opens the options when the user presses enter', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'Enter',
          key: 'Enter',
        })
        const itemList = screen.getAllByRole('listitem')
        expect(itemList).toHaveLength(numberRange.length)
      })
      it('Opens the options when the user presses space', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'Space',
          key: ' ',
        })
        const itemList = screen.getAllByRole('listitem')
        expect(itemList).toHaveLength(numberRange.length)
      })

      it('Opens the options when the user presses arrow down', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'ArrowDown',
          key: 'ArrowDown',
        })
        const itemList = screen.getAllByRole('listitem')
        expect(itemList).toHaveLength(numberRange.length)
      })
    })

    describe('When the select is open and the list of options is visible', () => {
      it('is navigable with the tab key', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'Enter',
          key: 'Enter',
        })
        const secondItem = screen.getByText('value 2')
        spyOn(secondItem, 'focus')

        expect(secondItem.focus).not.toHaveBeenCalled()
        userEvent.tab()
        expect(secondItem.focus).toHaveBeenCalled()
      })

      it('is navigable with the arrow keys', () => {
        render(<Select options={numberRange} onChange={() => true} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'Enter',
          key: 'Enter',
        })
        const secondItem = screen.getByText('value 2')
        spyOn(secondItem, 'focus')

        const thirdItem = screen.getByText('value 3')
        spyOn(thirdItem, 'focus')

        expect(secondItem.focus).not.toHaveBeenCalled()

        fireEvent.keyUp(secondItem, {
          code: 'ArrowDown',
          key: 'ArrowDown',
        })
        expect(secondItem.focus).toHaveBeenCalledTimes(1)

        fireEvent.keyUp(secondItem, {
          code: 'ArrowDown',
          key: 'ArrowDown',
        })
        expect(thirdItem.focus).toHaveBeenCalledTimes(1)

        fireEvent.keyUp(secondItem, {
          code: 'ArrowUp',
          key: 'ArrowUp',
        })
        expect(secondItem.focus).toHaveBeenCalledTimes(2)
      })

      it('is selectable with the enter key', () => {
        const changeSpy = jest.fn()
        render(<Select options={numberRange} onChange={changeSpy} />)
        const clickElement = screen.getByText('value 1')
        fireEvent.keyUp(clickElement, {
          code: 'Enter',
          key: 'Enter',
        })
        const secondItem = screen.getByText('value 2')
        fireEvent.keyUp(secondItem, {
          code: 'Enter',
          key: 'Enter',
        })
        expect(changeSpy).toHaveBeenCalledWith(numberRange[1])
      })
    })
  })
})
