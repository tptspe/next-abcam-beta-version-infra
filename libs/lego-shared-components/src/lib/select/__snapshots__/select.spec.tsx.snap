// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The select element Snapshots Matches the snapshot on initial state 1`] = `
<div
  class="select"
>
  <div
    class="selectedValueContainer"
    role="button"
    tabindex="0"
  >
    value 1
    <svg
      aria-label="Open select"
    />
  </div>
  <div
    class="content"
    role="button"
  />
</div>
`;

exports[`The select element Snapshots Matches the snapshot when opened 1`] = `
<div
  class="select open"
>
  <div
    class="selectedValueContainer"
    role="button"
    tabindex="0"
  >
    value 1
    <svg
      aria-label="Open select"
    />
  </div>
  <div
    class="content"
    role="button"
  >
    <ul
      role="presentation"
    >
      <li
        aria-label="value 1"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 1
        </div>
      </li>
      <li
        aria-label="value 2"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 2
        </div>
      </li>
      <li
        aria-label="value 3"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 3
        </div>
      </li>
      <li
        aria-label="value 4"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 4
        </div>
      </li>
      <li
        aria-label="value 5"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 5
        </div>
      </li>
      <li
        aria-label="value 6"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 6
        </div>
      </li>
      <li
        aria-label="value 7"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 7
        </div>
      </li>
      <li
        aria-label="value 8"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 8
        </div>
      </li>
      <li
        aria-label="value 9"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 9
        </div>
      </li>
      <li
        aria-label="value 10"
      >
        <div
          role="button"
          tabindex="0"
        >
          value 10
        </div>
      </li>
    </ul>
  </div>
</div>
`;
